EXPLORATORY DATA ANALYSIS

Çok kullanılan metodlardan birisi olan, value_counts() metodu bir sütun içerisindeki her bir veriden kaç adet bulunduğu bilgisini verir. 

In [1]:
import pandas as pd

In [2]:
data = pd.read_csv("pokemon.csv")

In [3]:
print(data['Type 1'].value_counts(dropna =False))

Water       112
Normal       98
Grass        70
Bug          69
Psychic      57
Fire         52
Electric     44
Rock         44
Dragon       32
Ground       32
Ghost        32
Dark         31
Poison       28
Steel        27
Fighting     27
Ice          24
Fairy        17
Flying        4
Name: Type 1, dtype: int64


Çıktıyı inceleyelim. "Type 1" sütunu pokemonun türü hakkında bilgiler içerir. Mesela çıktının ilk satırında, 800 adet pokemondan 112 tanesinin su pokemonu olduğunu ifade eder.

İstatistikle ilgili bir kavram olan "outliers" ifadesinden bahsedelim. Bu kavram "ayrık, aykırı" anlamına gelmektedir.

Bir örnek üzerinden anlatalım.

1,4,5,6,8,9,11,12,13,14,15,16,17 şeklinde bir seri olsun. Median değeri serinin ortasındaki sayıdır. Serinin ortasındaki sayı 11 dir ve bu durumda median=11 dir. median kavramı "quartile 50%" şeklindede adlandırılıyor.

"lower quartile" kavramından bahsedelim. lower quartile, 1 ile 11 arasındaki median değeridir. Yani 1 ile 11 sayılarının ortasındaki sayıdır ve bu sayı 6'dır. lower quartile sayısı 6'dır. "Lower quartile" kavramına "Q1" de denir. Aynı zamanda "25% quartile" da denir. Yani "lower quartile" = Q1 = 25% quartile dır.

Birde "upper quartile" kavramından bahsedelim. 11 ile 17 arasındaki sayı upper quartile değeridir. Bu durumda upper quartile=14 dür. "upper quartile" kavramına "Q3" de denir. Aynı zamanda "75% quartile" da denir. 

Yani özetle:

Minimum ------------ Lower Quartile(Q1) ------------ Median ------------ Upper Quartile(Q3) ------------ Maximum      

Sayıların ortasındaki Median'dır. Median ile en sağdaki sayının ortasındaki değer Upper Quartile 'dir. Median ile en soldaki sayının ortasındaki değer Lower Quartile 'dir. 

Outlier: Bir istatistiğe göre bir data'daki diğer data'lardan aykırı olan yani çok yüksek ya da çok düşük olan data'dır. 

Mesela 1000 tane ev olsun. Bu evlerden en ucuzu 100 dolar, en pahalısı 1000 dolardır. 1001. evin fiyatını 1.000.000.000 dolar olarak belirlersek bu bir outlier'dir. Çünkü diğer data'lardan çok yukarıda bir değer. Bu yüzden bu bir aykırı sayıdır. 

Peki ne kadar yukarıda olduğu zaman aykırı sayı olarak nitelendirebiliriz?:

Q3 ile Q1 arasındaki mesafeyi "IQR" olarak adlandırırız. Yani IQR=Q3-Q1 dir. IQR değerini 1.5 ile çarpıp, Q1 den çıkarırız ve elde edilen değerin altındakilere outlier deriz. Yani "Q1 - 1.5(Q3-Q1)" ifadesinden elde edilen değerin altındaki değerlere outlier deriz. 1.5*IQR değeri Lower Quartile ile Minimum arasındaki mesafedir. Q1 den 1.5*IQR değerini çıkarırsak Mimimum değerinin altındaki bir noktayı elde ederiz. Elde edilen bu nokta outlier değeridir. 

Veya Q3 ile Maximum arasındaki mesafe 1.5*IQR değeridir. Q3+1.5*IQR ifadesinin sonucunda Maximum değerinin üstündeki bir noktayı elde ederiz. Elde edilen bu nokta outlier değeridir. 

Yani "Q1 - 1.5(Q3-Q1)" ifadesinden elde edilen değerin altındaki değerler veya "Q3+1.5*IQR" fadesinden elde edilen değerin üstündeki değerler outlier değeridir. 

Describe() metodunu hatırlayın. Bu metod ile elde ettiğimiz çıktıya birde bu anlattıklarımızdan sonra bakalım:

We will use describe() method. Describe method includes:
count: number of entries
mean: average of entries
std: standart deviation
min: minimum entry
25%: first quantile
50%: median or second quantile
75%: third quantile
max: maximum entry

describe() metodunun çıktısı artık daha anlamlı :) 

Yukarıda yaptığımız hesapları describe() metodu bizim için yapıyor.

In [4]:
data.describe()

#          HP      Attack     Defense     Sp. Atk     Sp. Def  \
count  800.0000  800.000000  800.000000  800.000000  800.000000  800.000000   
mean   400.5000   69.258750   79.001250   73.842500   72.820000   71.902500   
std    231.0844   25.534669   32.457366   31.183501   32.722294   27.828916   
min      1.0000    1.000000    5.000000    5.000000   10.000000   20.000000   
25%    200.7500   50.000000   55.000000   50.000000   49.750000   50.000000   
50%    400.5000   65.000000   75.000000   70.000000   65.000000   70.000000   
75%    600.2500   80.000000  100.000000   90.000000   95.000000   90.000000   
max    800.0000  255.000000  190.000000  230.000000  194.000000  230.000000   

            Speed  Generation  
count  800.000000   800.00000  
mean    68.277500     3.32375  
std     29.060474     1.66129  
min      5.000000     1.00000  
25%     45.000000     2.00000  
50%     65.000000     3.00000  
75%     90.000000     5.00000  
max    180.000000     6.00000

describe() metodu sadece numerik değerleri döndürür.

Çıktıyı inceleyelim.

HP sütunu pokemonların canını ifade eder. Pokemonların ortalama canı 69.25 dir. Minimum can yani canı en az olan pokemonun 1 tane canı varmış. Maksimum can yani canı en çok olan pokemonun 255 tane canı varmış. Pokemonların canının ortasındaki değer (50%) 65 dir. Lowe quartile değeri 50 dir. Upper quartile (75%) değeri 80 dir.

Peki neden mean gibi bir değer varken median değerinide hesaplamak istiyoruz?:

5 çalışanı olan bir şirket olsun. Çalışanların maaşları sırasıyla 1,1,1,1,1000 olsun. Çalışanların maaşlarına ortalama maaş değerine göre zam yapılıp yapılmaması durumu söz konusu. 1004/5=200.8 ortalama maaş değeridir. Maaş ortalamasına göre çalışanların yeterince maaş aldığı kararına varılmıştır ve zam yapılmamıştır. Çünkü mean yani ortalama bulmak bizi yanılttı. Eğer median değerini bulmuş olsaydık yani diziyi sıraladığımızda ortadaki değere göre zam yapıp yapmamaya karar verseydik, şirkette çalışanların %50 sinin maaşı 1'den az kanısına varırdık ve çalışanların maaşının az olduğu anlaşılarak maaşlara zam yapılırdı. Mean değerine göre karar verseydik, maaşı 1000 olan çalışan karar verme konusunda yanıltacaktı. 

İşte bu yüzden Median, quartile 25% ve quartile 75% değerleri önemli istatistiksel değerleridir.